[docs][data grid] Revise the Row Grouping doc (#16217)
mui/mui-x: docs/data/data-grid/row-grouping/row-grouping.md

# Data Grid - Row grouping [<span class="plan-premium"></span>](/x/introduction/licensing/#premium-plan 'Premium plan')

<p class="description">Group rows together based on column values in the Data Grid.</p>

The Data Grid Premium provides a row grouping feature which makes it possible to create subsets of rows based on repeated column values or custom functions.

For example, in the demo below, movies are grouped based on their respective values in the production company column:

{{"demo": "RowGroupingBasicExample.js", "bg": "inline", "defaultCodeOpen": false}}

:::info
This document covers client-side implementation.

> For row grouping on the server side, see [server-side row grouping](/x/react-data-grid/server-side-data/row-grouping/).

Author: This Vale warning is incorrect 😭 I'll open a separate PR to address it

<img width="1224" alt="Screenshot 2025-02-28 at 9 57 48 AM" src="https://github.com/user-attachments/assets/8dd5a5ee-a5be-4a81-8627-bb4bd78dc35d" />

The demo below uses `leafField` to display values from the **Title** column in the leaves.

{{"demo": "RowGroupingLeafWithValue.js", "bg": "inline", "defaultCodeOpen": false}}

### Hide descendant count

By default, row group cells display the number of descendant rows they contain, in parentheses.
Use the `hideDescendantCount` property of the `groupingColDef` to hide this number.

{{"demo": "RowGroupingHideDescendantCount.js", "bg": "inline", "defaultCodeOpen": false}}

### Hide grouped columns

By default, [grouped columns](#grouping-columns) remain visible when grouping is applied.
This means that when there's only one grouping criterion, the grouping column and the grouped column both display the same values, which may be redundant or unnecessary.
You can use the `useKeepGroupedColumnsHidden` utility hook to hide grouped columns.
This automatically hides grouped columns when added to the model, and displays them when removed.

:::warning
This hook is not compatible with the deprecated column property `hide`.

You can manage column visibility with `columnVisibilityModel`, `initialState`, or both together.
To do so, pass them to the hook parameters as shown below.

> :::

Reviewer: This property (`GridColDef['hide']`) is long [gone](https://github.com/mui/mui-x/pull/6505/files#diff-654c76d2458fcc7257077878eb1d07bbc743abae9c219ddb3c9296f1a50ffa36R61). 😄 
Not sure why it's still mentioned. I'd suggest to remove this callout completely.

> :::warning
> This hook is not compatible with the deprecated column property `hide`.
> 
> You can manage column visibility with `columnVisibilityModel`, `initialState`, or both together.
> To do so, pass them to the hook parameters as shown below.
> :::

Author: ```suggestion
```

{{"demo": "RowGroupingUseKeepGroupedColumnsHidden.js", "bg": "inline", "defaultCodeOpen": false}}

## Disable row grouping

### For all columns

To disable row grouping for all columns, set the `disableRowGrouping` prop to `true`.

> This disables all features related to row grouping, even if a model is provided.

Reviewer: This statement seems to be wrong, as mentioned in the section below (grouping non..... programatically)
```suggestion
This disables all features related to row grouping, unless a model is provided, in which case the row grouping works in a read-only mode as detailed in [this](#grouping-non-groupable-columns) section.
```

Author: ```suggestion
This disables all features related to row grouping, unless a model is provided, in which case row grouping is set to read-only mode as described in [Grouping non-groupable columns programmatically](#grouping-non-groupable-columns-programmatically).
```

{{"demo": "RowGroupingDisabled.js", "bg": "inline", "defaultCodeOpen": false}}

### For specific columns

To disable grouping for a specific column, set the `groupable` property on its `GridColDef` to `false`.

In the example below, the **Director** column cannot be grouped, even though there are repeating values.
(The **Title** and **Gross** columns cannot be grouped in any examples in this doc because there are no repeating values.)

{{"demo": "RowGroupingColDefCanBeGrouped.js", "bg": "inline", "defaultCodeOpen": false}}

> ### Grouping non-groupable columns

Reviewer: I'd keep "programmatically" for the same reason as https://github.com/mui/mui-x/pull/16217#discussion_r2003721041

Author: ```suggestion
### Grouping non-groupable columns programmatically
```

> To apply row grouping programmatically on non-groupable columns (columns with `groupable: false` in the `GridColDef`), you can provide the row grouping model in one of three ways:

Reviewer: I find having a link to the reference API helpful. Just curious if the rationale for removing it is to avoid potential misdirection?

Author: It seemed redundant and maybe a little arbitrary to link to it here since the column definition is mentioned but not linked in other places throughout the doc. If it's specifically useful in this context, then maybe we should be more explicit about calling it out as a resource here. Since this is immediately followed by a list, how about an info or success callout below that directing the reader to the API reference for more complete details?

In the example below, the **Director** column cannot be grouped, even though there are repeating values.
(The **Title** and **Gross** columns cannot be grouped in any examples in this doc because there are no repeating values.)

{{"demo": "RowGroupingColDefCanBeGrouped.js", "bg": "inline", "defaultCodeOpen": false}}

### Grouping non-groupable columns

To apply row grouping programmatically on non-groupable columns (columns with `groupable: false` in the `GridColDef`), you can provide the row grouping model in one of three ways:

1. Pass `rowGrouping.model` to the `initialState` prop. This [initializes grouping](/x/react-data-grid/row-grouping/#initializing-row-grouping) with the provided model.
2. Provide the `rowGroupingModel` prop. This [controls grouping](/x/react-data-grid/row-grouping/#controlled-row-grouping) with the provided model.
3. Call the API method `setRowGroupingModel`. This sets the aggregation with the provided model.

In the following example, the **Company** column is not groupable through the interface, but the `rowGroupingModel` prop is passed to generate a read-only row group.

{{"demo": "RowGroupingReadOnly.js", "bg": "inline", "defaultCodeOpen": false}}

> ## Complex grouping values

Reviewer: Debatable, but one benefit of more detailed headings like the previous one could be to get nice search results on Algolia:

<img width="660" alt="image" src="https://github.com/user-attachments/assets/a9bb5de4-f8a1-4993-917f-a9065f428f3c" />

Author: Good call, I'm usually trying to optimize for quickly scanning the table of contents nav when I make changes like this, but search is important to factor in as well. I'll revert this one and reevaluate the other headers here.

And I always add redirects when page slugs change, but it never occurred to me to do it for header links too! 😱

Reviewer 2: Oh, I was wrong about the redirects for anchors (at least server-side redirects) since the hash is not sent to the server.
You can redirect **to** URL with a hash, but you can't redirect **from** URL with a hash, and this is where my confusion came from – sorry about that! 😅

If your column also has a `valueGetter` property, the value passed to the `groupingValueGetter` method will still be the row value from the `row[field]`.
:::

> ### Rows with missing groups

Reviewer: It doesn't seem to be a sub-section of `groupingValueGetter`. Wdyt?
```suggestion
## Rows with missing groups
```

Author: It seemed to follow from the h2 above since that section mentions that the grouping value must be a string, a number, `null`, or `undefined`, and then this section covers when it's `null` or `undefined`. But now that I'm reading it again, I'm realizing that the section above talks about "grouping values" while this section is about "grouping keys of grouping criteria." Are these different things? I might've gotten confused.

If a grouping criterion's key is `null` or `undefined` for a given row, the Data Grid will treat that row as if it doesn't have a value and exclude it from grouping.

The demo below illustrates this behavior—movies are grouped by **Cinematic Universe**, and rows with no value for this column are displayed individually before those with values are displayed in groups:

{{"demo": "RowGroupingRowsWithMissingGroups.js", "bg": "inline", "defaultCodeOpen": false}}

## Group expansion

> By default, all groups are initially displayed collapsed.

Reviewer: I might be wrong but `by default` and `initially` seem to convey a similar meaning.
```suggestion
Initially, all groups are displayed collapsed.
```

Author: Unless I've misunderstood, I believe they're conveying two slightly different things: "By default" says this is how the Data Grid is set up to behave out of out the box, whereas "initially" clarifies that this is what happens when data is first loaded (it's displayed collapsed). So it makes sense to use both in this sentence.

If your column also has a `valueGetter` property, the value passed to the `groupingValueGetter` method will still be the row value from the `row[field]`.
:::

### Rows with missing groups

If a grouping criterion's key is `null` or `undefined` for a given row, the Data Grid will treat that row as if it doesn't have a value and exclude it from grouping.

The demo below illustrates this behavior—movies are grouped by **Cinematic Universe**, and rows with no value for this column are displayed individually before those with values are displayed in groups:

{{"demo": "RowGroupingRowsWithMissingGroups.js", "bg": "inline", "defaultCodeOpen": false}}

> ## Expanding groups on first load

Reviewer: I'd revert this change since it's not necessarily on the first load, it technically happens whenever the rows pre-processor runs which could happen multiple times during the lifecycle of a Data Grid.

Author: thanks for clarifying!

To expand the whole tree, set `defaultGroupingExpansionDepth = -1`.

{{"demo": "RowGroupingDefaultExpansionDepth.js", "bg": "inline", "defaultCodeOpen": false}}

Use the `isGroupExpandedByDefault()` prop to expand groups by default according to more complex logic. This prop is a callback that receives a node as an argument.
When defined, this callback always takes priority over the `defaultGroupingExpansionDepth` prop.

```tsx
isGroupExpandedByDefault={
  node => node.groupingField === 'company' && node.groupingKey === '20th Century Fox'
}
```

The example below uses this pattern to only expand the **20th Century Fox** group on first load:

{{"demo": "RowGroupingIsGroupExpandedByDefault.js", "bg": "inline", "defaultCodeOpen": false}}

Use the `setRowChildrenExpansion()` method on [the `apiRef` object](#apiref) to programmatically set the expansion of a row.
Changing the expansion of a row emits a `rowExpansionChange` event that you can listen for to react to the expansion change.

The demo below uses this pattern to implement the expansion toggle button:

{{"demo": "RowGroupingSetChildrenExpansion.js", "bg": "inline", "defaultCodeOpen": false}}

:::warning

> The `apiRef.current.setRowChildrenExpansion()` method is not compatible with [server-side tree data](/x/react-data-grid/server-side-data/tree-data/) or [server-side row grouping](/x/react-data-grid/server-side-data/row-grouping/).

Reviewer: ```suggestion
The `apiRef.current.setRowChildrenExpansion()` method is not compatible with [Server-side data-Tree data](/x/react-data-grid/server-side-data/tree-data/) or [Server-side data-Row grouping](/x/react-data-grid/server-side-data/row-grouping/).
```

If you're rendering leaves in one of those columns with the `leafField` property of `groupingColDef`, then sorting and filtering will be applied on the leaves for this grouping column based on the `sortComparator` and `filterOperators` of their original column.

If you want to render leaves but apply the sorting and filtering on the grouping criteria of the column, you can force it by setting the `mainGroupingCriteria` property `groupingColDef` to be equal to the grouping criteria.

In the example below, sorting and filtering from the **Company** grouping column are applied to the `company` field, while sorting and filtering from the **Director** grouping column are applied to the `director` field even though it has leaves.

{{"demo": "RowGroupingSortingMultipleGroupingColDef.js", "bg": "inline", "defaultCodeOpen": false}}

:::warning
If you're dynamically switching the `leafField` or `mainGroupingCriteria`, sorting and filtering models will not be cleaned up automatically, and the sorting or filtering will not be reapplied.
:::

> ## Automatic parent and children selection

Reviewer: Consider using a consistent singular or plural form for both—either “parents” and “children” or “parent” and “child.” I’d suggest the plural version. What do you think?

Author: Good call. My instinct would be to keep it singular: I can't put my finger on why exactly, but "automatic parent and child selection" reads better to me than "automatic parents and children selection."

Reviewer: Sure, let's go with it then. 👍

};
```

> ### Row selection propagation

Reviewer: I'd keep this under the same top level heading to avoid confusion about this something different than "automatic p.... selection"
```suggestion
```

When `rowSelectionPropagation.descendants` is set to `true`:

- Selecting a parent selects all of its filtered descendants automatically
- Deselecting a parent row deselects of all its filtered descendants automatically

When `rowSelectionPropagation.parents` is set to `true`:

- Selecting all the filtered descendants of a parent also selects the parent automatically
- Deselecting a descendant of a selected parent also deselects the parent automatically

The example below demonstrates the usage of the `rowSelectionPropagation` prop—use the checkboxes at the top to see how the selection behavior changes between parents and children.

{{"demo": "RowGroupingPropagateSelection.js", "bg": "inline", "defaultCodeOpen": false}}

Row selection propagation also affects the **Select all** checkbox like any other checkbox group.
Selected rows that do not pass the filtering criteria are automatically deselected when the filter is applied.
Row selection propagation is not applied to the unfiltered rows.

:::warning
Row selection propagation has some limitations:

- If `props.disableMultipleRowSelection` is set to `true`, then row selection propagation won't apply.

> - Row selection propagation is a client-side feature and does not support [server-side data sources](/x/react-data-grid/server-side-data/).

Reviewer: We never use plural of data source elsewhere, could be confused with the actual server side data sources rather than Data Grid's "data source" feature.
```suggestion
- Row selection propagation is a client-side feature and does not support [server-side data](/x/react-data-grid/server-side-data/).
```

{{"demo": "RowGroupingGetRowGroupChildren.js", "bg": "inline", "defaultCodeOpen": false}}

:::warning

> The `apiRef.current.getRowGroupChildren` method is not compatible with [server-side row grouping](/x/react-data-grid/server-side-data/row-grouping/) because not all rows may be available to retrieve in any given instance.

Reviewer: ```suggestion
The `apiRef.current.getRowGroupChildren()` method is not compatible with [Server-side data-Row grouping](/x/react-data-grid/server-side-data/row-grouping/) because not all rows may be available to retrieve in any given instance.
```

With the row group panel, users would be able to control which columns are used for grouping by dragging them inside the panel.

## Advanced use cases

The demo below provides an example of row grouping that's closer to a real-world use case for this feature, grouping a large dataset of contacts based on the types of goods they provide (via the hidden **Commodity** column).

> See [Row grouping recipes](/x/react-data-grid/recipes-row-grouping/) for more advanced use cases.

Reviewer: Would it make more sense to move this line after the demo?
